File: containers/heartbeat/src/client/telemetry.go
package client

// IMPORTS
import (
	"bytes"
	"encoding/json"
	"fmt"
	"net/http"
	"time"
)

// TYPES

// TelemetryClient sends log entries to the centralized telemetry service
type TelemetryClient struct {
	url    string
	client *http.Client
}

// LogRequest represents a telemetry log entry with correlation ID for distributed tracing
type LogRequest struct {
	Level   string `json:"level"`
	Message string `json:"message"`
	LogID   string `json:"log_id"`
	Source  string `json:"source"`
}

// FUNCTIONS

// NewTelemetryClient creates a client for sending logs to telemetry service
func NewTelemetryClient(url string, timeout time.Duration) *TelemetryClient {
	const minTimeout = 1 * time.Second
	const maxTimeout = 60 * time.Second
	if timeout < minTimeout {
		timeout = minTimeout
	} else if timeout > maxTimeout {
		timeout = maxTimeout
	}
	return &TelemetryClient{
		url: url,
		client: &http.Client{
			Timeout: timeout,
		},
	}
}

// Log sends a log entry to telemetry with correlation ID for request tracing
func (c *TelemetryClient) Log(level, message, logID string) error {
	req := LogRequest{
		Level:   level,
		Message: message,
		LogID:   logID,
		Source:  "heartbeat",
	}

	body, err := json.Marshal(req)
	if err != nil {
		return fmt.Errorf("failed to marshal log request: %w", err)
	}

	resp, err := c.client.Post(c.url, "application/json", bytes.NewReader(body))
	if err != nil {
		return fmt.Errorf("failed to send log: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return fmt.Errorf("telemetry returned status %d", resp.StatusCode)
	}

	return nil
}

File: containers/webservice/src/lib/telemetry.ts
const TELEMETRY_HOST = process.env.TELEMETRY_HOST || "telemetry";
const TELEMETRY_PORT = process.env.TELEMETRY_PORT || "8081";
const TELEMETRY_URL = `http://${TELEMETRY_HOST}:${TELEMETRY_PORT}/log`;

interface LogRequest {
  level: string;
  message: string;
  log_id?: string;
  source: string;
}

export async function log(
  level: "debug" | "info" | "warn" | "error",
  message: string,
  logId?: string,
): Promise<void> {
  const request: LogRequest = {
    level,
    message,
    log_id: logId,
    source: "webservice",
  };

  try {
    await fetch(TELEMETRY_URL, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(request),
    });
  } catch (error) {
    console.error("Failed to send log to telemetry:", error);
  }
}

export function getLogIdFromHeaders(headers: Headers): string | undefined {
  return headers.get("x-log-id") || undefined;
}
